File: python/src/context_firewall/classifier.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation

"""
Keyword-based data classifier for context-firewall.

FIRE LINE: Classification uses KEYWORD MATCHING only.
No ML, no LLM, no embeddings, no external API calls.
The classifier is fully deterministic and reproducible without GPU resources.
"""

from __future__ import annotations

from .types import DataClassification, DataPayload, KeywordRule

__all__ = [
    "DataClassifier",
    "DataClassifierOptions",
    "DEFAULT_KEYWORD_RULES",
]

# ---------------------------------------------------------------------------
# Default keyword rules
# ---------------------------------------------------------------------------

DEFAULT_KEYWORD_RULES: tuple[KeywordRule, ...] = (
    # --- Health domain ---
    KeywordRule(
        id="health-medical-general",
        domain="health",
        data_type="medical",
        keywords=[
            "patient", "doctor", "physician", "hospital", "clinic",
            "medical", "medicine", "treatment", "symptoms", "symptom",
            "illness", "disease", "condition", "surgery", "procedure",
            "referral", "immunization", "vaccination",
        ],
    ),
    KeywordRule(
        id="health-prescription",
        domain="health",
        data_type="prescription",
        keywords=[
            "prescription", "medication", "dosage", "mg", "tablet",
            "capsule", "pharmacy", "pharmacist", "refill", "drug",
            "antibiotic", "insulin", "inhaler", "ointment",
        ],
    ),
    KeywordRule(
        id="health-diagnosis",
        domain="health",
        data_type="diagnosis",
        keywords=[
            "diagnosis", "diagnosed", "prognosis", "icd", "icd-10", "icd-11",
            "cancer", "diabetes", "hypertension", "asthma", "allergy",
            "chronic", "acute", "disorder", "syndrome", "infection",
        ],
    ),
    KeywordRule(
        id="health-mental-health",
        domain="health",
        data_type="mental-health",
        keywords=[
            "therapy", "therapist", "psychiatrist", "psychologist",
            "counseling", "mental-health", "depression", "anxiety",
            "bipolar", "adhd", "ptsd", "ocd", "schizophrenia", "psychosis",
            "antidepressant", "anxiolytic", "ssri",
        ],
    ),
    KeywordRule(
        id="health-laboratory",
        domain="health",
        data_type="laboratory",
        keywords=[
            "lab-result", "blood-test", "urine-test", "biopsy",
            "pathology", "mri", "ct-scan", "x-ray", "ultrasound",
            "ecg", "ekg", "cholesterol", "glucose", "hemoglobin",
            "white-blood-cell", "red-blood-cell", "creatinine",
        ],
    ),
    KeywordRule(
        id="health-clinical",
        domain="health",
        data_type="clinical",
        keywords=[
            "ehr", "emr", "fhir", "hl7", "clinical-note",
            "discharge-summary", "nursing", "nurse", "ward", "icu",
            "emergency-room", "urgent-care", "inpatient", "outpatient",
            "telemedicine", "telehealth",
        ],
    ),
    # --- Financial domain ---
    KeywordRule(
        id="financial-banking",
        domain="financial",
        data_type="banking",
        keywords=[
            "bank", "account-number", "routing-number", "iban", "swift",
            "wire-transfer", "deposit", "withdrawal", "overdraft",
            "savings", "checking", "balance", "transaction", "statement",
        ],
    ),
    KeywordRule(
        id="financial-credit-card",
        domain="financial",
        data_type="credit-card",
        keywords=[
            "credit-card", "debit-card", "card-number", "cvv", "expiry",
            "visa", "mastercard", "amex", "american-express", "discover",
            "payment-card", "cardholder",
        ],
    ),
    KeywordRule(
        id="financial-tax",
        domain="financial",
        data_type="tax",
        keywords=[
            "tax-return", "irs", "hmrc", "w-2", "1099", "ein", "tin",
            "ssn", "social-security", "deduction", "filing",
            "taxable-income", "refund", "audit", "capital-gains",
        ],
    ),
    KeywordRule(
        id="financial-investment",
        domain="financial",
        data_type="investment",
        keywords=[
            "portfolio", "stock", "equity", "bond", "etf", "mutual-fund",
            "brokerage", "dividend", "401k", "ira", "roth", "ticker",
            "securities", "cryptocurrency", "bitcoin", "ethereum",
        ],
    ),
    KeywordRule(
        id="financial-salary",
        domain="financial",
        data_type="salary",
        keywords=[
            "salary", "payroll", "payslip", "wage", "compensation",
            "bonus", "commission", "income", "net-pay", "gross-pay",
            "pension", "401k-contribution",
        ],
    ),
    # --- Personal domain ---
    KeywordRule(
        id="personal-family",
        domain="personal",
        data_type="family",
        keywords=[
            "family", "spouse", "partner", "husband", "wife", "child",
            "children", "son", "daughter", "parent", "mother", "father",
            "sibling", "brother", "sister", "grandparent", "grandchild",
            "relative",
        ],
    ),
    KeywordRule(
        id="personal-relationship",
        domain="personal",
        data_type="relationship",
        keywords=[
            "relationship", "romantic", "dating", "marriage", "divorce",
            "engaged", "breakup", "intimate", "personal-life",
            "friend", "friendship",
        ],
    ),
    KeywordRule(
        id="personal-home-address",
        domain="personal",
        data_type="home-address",
        keywords=[
            "home-address", "residential", "street-address", "zip-code",
            "postal-code", "neighborhood", "apartment", "house",
            "home-phone", "home-email",
        ],
    ),
    KeywordRule(
        id="personal-contact",
        domain="personal",
        data_type="personal-contact",
        keywords=[
            "personal-email", "cell-phone", "personal-phone",
            "home-contact", "next-of-kin", "emergency-contact",
        ],
    ),
    # --- Work domain ---
    KeywordRule(
        id="work-professional",
        domain="work",
        data_type="professional",
        keywords=[
            "meeting", "agenda", "deadline", "project", "deliverable",
            "stakeholder", "client", "vendor", "invoice", "contract",
            "proposal", "presentation", "report", "sprint", "milestone",
            "kpi", "okr", "quarterly", "annual-review",
        ],
    ),
    KeywordRule(
        id="work-communication",
        domain="work",
        data_type="work-communication",
        keywords=[
            "slack", "teams", "email-thread", "work-email", "corporate",
            "colleagues", "manager", "employee", "hr", "human-resources",
            "onboarding", "offboarding", "performance-review",
        ],
    ),
)


# ---------------------------------------------------------------------------
# Classifier options
# ---------------------------------------------------------------------------


class DataClassifierOptions:
    """
    Configuration options for :class:`DataClassifier`.

    :param additional_rules: Extra keyword rules merged with (or replacing) built-in defaults.
    :param replace_default_rules: When ``True``, built-in rules are discarded.
    :param fallback_domain: Domain returned when no keywords match. Default: ``"work"``.
    """

    def __init__(
        self,
        additional_rules: list[KeywordRule] | None = None,
        replace_default_rules: bool = False,
        fallback_domain: str = "work",
    ) -> None:
        self.additional_rules: list[KeywordRule] = additional_rules or []
        self.replace_default_rules: bool = replace_default_rules
        self.fallback_domain: str = fallback_domain


# ---------------------------------------------------------------------------
# DataClassifier class
# ---------------------------------------------------------------------------


class DataClassifier:
    """
    Keyword-based data classifier.

    Scans a :data:`~context_firewall.types.DataPayload` for known keywords and
    returns the domain and data type(s) that best match. Classification is:

    - **Deterministic**: same input always produces same output.
    - **Auditable**: matched keywords are reported in the result.
    - **Transparent**: no model weights, no external API calls.

    FIRE LINE: This class must never call an LLM or use embedding-based similarity.
    """

    def __init__(self, options: DataClassifierOptions | None = None) -> None:
        opts = options or DataClassifierOptions()
        if opts.replace_default_rules:
            self._rules: tuple[KeywordRule, ...] = tuple(opts.additional_rules)
        else:
            self._rules = DEFAULT_KEYWORD_RULES + tuple(opts.additional_rules)
        self._fallback_domain: str = opts.fallback_domain

    def classify(self, data: DataPayload) -> DataClassification:
        """
        Classify a data payload by scanning all string values for keyword matches.

        The winning domain is the one with the most keyword hits across all its rules.
        Confidence is the ratio of matched keywords to total unique keywords in the
        winning domain's rules, capped at 1.0.

        :param data: The data payload to classify.
        :returns: A :class:`~context_firewall.types.DataClassification` result.
        """
        text_content = self._extract_text_content(data)
        normalised = text_content.lower()

        domain_hits: dict[str, int] = {}
        detected_types: set[str] = set()
        all_matched_keywords: list[str] = []

        for rule in self._rules:
            rule_matches: list[str] = []
            for keyword in rule.keywords:
                if self._keyword_matches(normalised, keyword.lower()):
                    rule_matches.append(keyword)

            if rule_matches:
                domain_hits[rule.domain] = domain_hits.get(rule.domain, 0) + len(rule_matches)
                detected_types.add(rule.data_type)
                all_matched_keywords.extend(rule_matches)

        if not domain_hits:
            return DataClassification(
                domain=self._fallback_domain,
                detected_types=[],
                matched_keywords=[],
                confidence=0.0,
            )

        winning_domain = max(domain_hits, key=lambda d: domain_hits[d])
        highest_hits = domain_hits[winning_domain]
        winning_keyword_count = self._count_keywords_for_domain(winning_domain)
        confidence = min(
            1.0,
            highest_hits / winning_keyword_count if winning_keyword_count > 0 else 0.0,
        )

        return DataClassification(
            domain=winning_domain,
            detected_types=list(detected_types),
            matched_keywords=list(set(all_matched_keywords)),
            confidence=confidence,
        )

    def get_rules(self) -> tuple[KeywordRule, ...]:
        """Return all active keyword rules (built-in + custom)."""
        return self._rules

    # -----------------------------------------------------------------------
    # Private helpers
    # -----------------------------------------------------------------------

    def _extract_text_content(self, data: DataPayload) -> str:
        parts: list[str] = []
        self._collect_strings(data, parts)
        return " ".join(parts)

    def _collect_strings(self, value: object, parts: list[str]) -> None:
        if isinstance(value, str):
            parts.append(value)
        elif isinstance(value, list):
            for item in value:
                self._collect_strings(item, parts)
        elif isinstance(value, dict):
            for val in value.values():
                self._collect_strings(val, parts)
        # int, float, bool, None — skip

    def _keyword_matches(self, normalised_text: str, keyword: str) -> bool:
        """
        Check whether *keyword* appears in *normalised_text*.
        Also checks with hyphens replaced by spaces for compound terms.
        """
        if self._contains_whole_word(normalised_text, keyword):
            return True
        spaced = keyword.replace("-", " ")
        if spaced != keyword and self._contains_whole_word(normalised_text, spaced):
            return True
        return False

    @staticmethod
    def _contains_whole_word(text: str, term: str) -> bool:
        """
        Check that *term* appears in *text* as a whole word.
        Word characters are ``[a-z0-9\\-_]``.
        """
        idx = text.find(term)
        if idx == -1:
            return False

        before = text[idx - 1] if idx > 0 else " "
        after_idx = idx + len(term)
        after = text[after_idx] if after_idx < len(text) else " "

        def is_word_char(char: str) -> bool:
            return char.isalnum() or char in ("-", "_")

        return not is_word_char(before) and not is_word_char(after)

    def _count_keywords_for_domain(self, domain: str) -> int:
        keywords: set[str] = set()
        for rule in self._rules:
            if rule.domain == domain:
                keywords.update(kw.lower() for kw in rule.keywords)
        return len(keywords)

File: typescript/src/classifier.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

/**
 * @module classifier
 * Keyword-based data classifier for context-firewall.
 *
 * FIRE LINE: Classification uses KEYWORD MATCHING only.
 * No ML, no LLM, no embeddings, no external API calls.
 * The classifier is fully deterministic and reproducible without GPU resources.
 */

import {
  DataClassification,
  DataPayload,
  KeywordRule,
  KeywordRuleSchema,
} from "./types.js";

// ---------------------------------------------------------------------------
// Default keyword rules
// ---------------------------------------------------------------------------

/**
 * Built-in keyword rules covering the four default domains.
 * Each rule maps a set of domain-specific keywords to a data type category.
 */
const DEFAULT_KEYWORD_RULES: ReadonlyArray<KeywordRule> = [
  // --- Health domain ---
  {
    id: "health-medical-general",
    domain: "health",
    dataType: "medical",
    keywords: [
      "patient",
      "doctor",
      "physician",
      "hospital",
      "clinic",
      "medical",
      "medicine",
      "treatment",
      "symptoms",
      "symptom",
      "illness",
      "disease",
      "condition",
      "surgery",
      "procedure",
      "referral",
      "appointment-health",
      "immunization",
      "vaccination",
    ],
  },
  {
    id: "health-prescription",
    domain: "health",
    dataType: "prescription",
    keywords: [
      "prescription",
      "medication",
      "dosage",
      "mg",
      "tablet",
      "capsule",
      "pharmacy",
      "pharmacist",
      "refill",
      "drug",
      "antibiotic",
      "insulin",
      "inhaler",
      "ointment",
    ],
  },
  {
    id: "health-diagnosis",
    domain: "health",
    dataType: "diagnosis",
    keywords: [
      "diagnosis",
      "diagnosed",
      "prognosis",
      "icd",
      "icd-10",
      "icd-11",
      "cancer",
      "diabetes",
      "hypertension",
      "asthma",
      "allergy",
      "chronic",
      "acute",
      "disorder",
      "syndrome",
      "infection",
    ],
  },
  {
    id: "health-mental-health",
    domain: "health",
    dataType: "mental-health",
    keywords: [
      "therapy",
      "therapist",
      "psychiatrist",
      "psychologist",
      "counseling",
      "mental-health",
      "depression",
      "anxiety",
      "bipolar",
      "adhd",
      "ptsd",
      "ocd",
      "schizophrenia",
      "psychosis",
      "antidepressant",
      "anxiolytic",
      "ssri",
      "session-therapy",
    ],
  },
  {
    id: "health-laboratory",
    domain: "health",
    dataType: "laboratory",
    keywords: [
      "lab-result",
      "blood-test",
      "urine-test",
      "biopsy",
      "pathology",
      "mri",
      "ct-scan",
      "x-ray",
      "ultrasound",
      "ecg",
      "ekg",
      "cholesterol",
      "glucose",
      "hemoglobin",
      "white-blood-cell",
      "red-blood-cell",
      "creatinine",
    ],
  },
  {
    id: "health-clinical",
    domain: "health",
    dataType: "clinical",
    keywords: [
      "ehr",
      "emr",
      "fhir",
      "hl7",
      "clinical-note",
      "discharge-summary",
      "nursing",
      "nurse",
      "ward",
      "icu",
      "emergency-room",
      "urgent-care",
      "inpatient",
      "outpatient",
      "telemedicine",
      "telehealth",
    ],
  },

  // --- Financial domain ---
  {
    id: "financial-banking",
    domain: "financial",
    dataType: "banking",
    keywords: [
      "bank",
      "account-number",
      "routing-number",
      "iban",
      "swift",
      "wire-transfer",
      "deposit",
      "withdrawal",
      "overdraft",
      "savings",
      "checking",
      "balance",
      "transaction",
      "statement",
    ],
  },
  {
    id: "financial-credit-card",
    domain: "financial",
    dataType: "credit-card",
    keywords: [
      "credit-card",
      "debit-card",
      "card-number",
      "cvv",
      "expiry",
      "visa",
      "mastercard",
      "amex",
      "american-express",
      "discover",
      "payment-card",
      "cardholder",
    ],
  },
  {
    id: "financial-tax",
    domain: "financial",
    dataType: "tax",
    keywords: [
      "tax-return",
      "irs",
      "hmrc",
      "w-2",
      "1099",
      "ein",
      "tin",
      "ssn",
      "social-security",
      "deduction",
      "filing",
      "taxable-income",
      "refund",
      "audit",
      "capital-gains",
    ],
  },
  {
    id: "financial-investment",
    domain: "financial",
    dataType: "investment",
    keywords: [
      "portfolio",
      "stock",
      "equity",
      "bond",
      "etf",
      "mutual-fund",
      "brokerage",
      "dividend",
      "401k",
      "ira",
      "roth",
      "ticker",
      "securities",
      "cryptocurrency",
      "bitcoin",
      "ethereum",
    ],
  },
  {
    id: "financial-salary",
    domain: "financial",
    dataType: "salary",
    keywords: [
      "salary",
      "payroll",
      "payslip",
      "wage",
      "compensation",
      "bonus",
      "commission",
      "income",
      "net-pay",
      "gross-pay",
      "pension",
      "401k-contribution",
    ],
  },

  // --- Personal domain ---
  {
    id: "personal-family",
    domain: "personal",
    dataType: "family",
    keywords: [
      "family",
      "spouse",
      "partner",
      "husband",
      "wife",
      "child",
      "children",
      "son",
      "daughter",
      "parent",
      "mother",
      "father",
      "sibling",
      "brother",
      "sister",
      "grandparent",
      "grandchild",
      "relative",
    ],
  },
  {
    id: "personal-relationship",
    domain: "personal",
    dataType: "relationship",
    keywords: [
      "relationship",
      "romantic",
      "dating",
      "marriage",
      "divorce",
      "engaged",
      "breakup",
      "intimate",
      "personal-life",
      "friend",
      "friendship",
    ],
  },
  {
    id: "personal-home-address",
    domain: "personal",
    dataType: "home-address",
    keywords: [
      "home-address",
      "residential",
      "street-address",
      "zip-code",
      "postal-code",
      "neighborhood",
      "apartment",
      "house",
      "home-phone",
      "home-email",
    ],
  },
  {
    id: "personal-contact",
    domain: "personal",
    dataType: "personal-contact",
    keywords: [
      "personal-email",
      "cell-phone",
      "personal-phone",
      "home-contact",
      "next-of-kin",
      "emergency-contact",
    ],
  },

  // --- Work domain ---
  {
    id: "work-professional",
    domain: "work",
    dataType: "professional",
    keywords: [
      "meeting",
      "agenda",
      "deadline",
      "project",
      "deliverable",
      "stakeholder",
      "client",
      "vendor",
      "invoice",
      "contract",
      "proposal",
      "presentation",
      "report",
      "sprint",
      "milestone",
      "kpi",
      "okr",
      "quarterly",
      "annual-review",
    ],
  },
  {
    id: "work-communication",
    domain: "work",
    dataType: "work-communication",
    keywords: [
      "slack",
      "teams",
      "email-thread",
      "work-email",
      "corporate",
      "colleagues",
      "manager",
      "employee",
      "hr",
      "human-resources",
      "onboarding",
      "offboarding",
      "performance-review",
    ],
  },
];

// ---------------------------------------------------------------------------
// Classifier options
// ---------------------------------------------------------------------------

/**
 * Configuration options for `DataClassifier`.
 */
export interface DataClassifierOptions {
  /**
   * Additional keyword rules to merge with (or replace) the built-in defaults.
   * If provided alongside `replaceDefaultRules: true`, only these rules are used.
   */
  readonly additionalRules?: ReadonlyArray<KeywordRule>;
  /**
   * When `true`, the built-in keyword rules are discarded and only
   * `additionalRules` are applied.
   * @default false
   */
  readonly replaceDefaultRules?: boolean;
  /**
   * The domain name to return when no keywords match.
   * @default "work"
   */
  readonly fallbackDomain?: string;
}

// ---------------------------------------------------------------------------
// DataClassifier class
// ---------------------------------------------------------------------------

/**
 * Keyword-based data classifier.
 *
 * Scans a `DataPayload` for known keywords and returns the domain and
 * data type(s) that best match. Classification is:
 * - Deterministic: same input always produces same output
 * - Auditable: matched keywords are reported in the result
 * - Transparent: no model weights, no external API calls
 *
 * FIRE LINE: This class must never call an LLM or use embedding-based similarity.
 */
export class DataClassifier {
  private readonly rules: ReadonlyArray<KeywordRule>;
  private readonly fallbackDomain: string;

  constructor(options: DataClassifierOptions = {}) {
    const {
      additionalRules = [],
      replaceDefaultRules = false,
      fallbackDomain = "work",
    } = options;

    // Validate all user-supplied rules
    const validatedAdditional = additionalRules.map((rule) =>
      KeywordRuleSchema.parse(rule)
    );

    this.rules = replaceDefaultRules
      ? validatedAdditional
      : [...DEFAULT_KEYWORD_RULES, ...validatedAdditional];

    this.fallbackDomain = fallbackDomain;
  }

  /**
   * Classify a `DataPayload` by scanning all string values for keyword matches.
   *
   * The winning domain is the one with the most keyword hits across all its rules.
   * Confidence is the ratio of matched keywords to total unique keywords checked,
   * capped at 1.0.
   *
   * @param data - The data payload to classify
   * @returns A `DataClassification` result with domain, detected types, matched keywords, and confidence
   */
  classify(data: DataPayload): DataClassification {
    const textContent = this.extractTextContent(data);
    const normalizedText = textContent.toLowerCase();

    // Accumulate hits per domain and per data type
    const domainHits = new Map<string, number>();
    const detectedTypeSet = new Set<string>();
    const allMatchedKeywords: string[] = [];

    for (const rule of this.rules) {
      const ruleMatches: string[] = [];

      for (const keyword of rule.keywords) {
        if (this.keywordMatches(normalizedText, keyword.toLowerCase())) {
          ruleMatches.push(keyword);
        }
      }

      if (ruleMatches.length > 0) {
        const currentHits = domainHits.get(rule.domain) ?? 0;
        domainHits.set(rule.domain, currentHits + ruleMatches.length);
        detectedTypeSet.add(rule.dataType);
        allMatchedKeywords.push(...ruleMatches);
      }
    }

    if (domainHits.size === 0) {
      return {
        domain: this.fallbackDomain,
        detectedTypes: [],
        matchedKeywords: [],
        confidence: 0,
      };
    }

    // Select the domain with the highest hit count
    let winningDomain = this.fallbackDomain;
    let highestHits = 0;

    for (const [domain, hits] of domainHits) {
      if (hits > highestHits) {
        highestHits = hits;
        winningDomain = domain;
      }
    }

    // Compute confidence: matched keywords / total keywords in winning domain's rules
    const winningDomainKeywordCount = this.countKeywordsForDomain(
      winningDomain
    );
    const confidence = Math.min(
      1,
      winningDomainKeywordCount > 0 ? highestHits / winningDomainKeywordCount : 0
    );

    return {
      domain: winningDomain,
      detectedTypes: Array.from(detectedTypeSet),
      matchedKeywords: Array.from(new Set(allMatchedKeywords)),
      confidence,
    };
  }

  /**
   * Return all active keyword rules (built-in + custom).
   */
  getRules(): ReadonlyArray<KeywordRule> {
    return this.rules;
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  /**
   * Extract all string values from a data payload, recursively flattening
   * nested objects and arrays. No ML is used here — pure string extraction.
   */
  private extractTextContent(data: DataPayload): string {
    const parts: string[] = [];
    this.collectStrings(data, parts);
    return parts.join(" ");
  }

  private collectStrings(value: unknown, parts: string[]): void {
    if (typeof value === "string") {
      parts.push(value);
    } else if (Array.isArray(value)) {
      for (const item of value) {
        this.collectStrings(item, parts);
      }
    } else if (value !== null && typeof value === "object") {
      for (const val of Object.values(value as Record<string, unknown>)) {
        this.collectStrings(val, parts);
      }
    }
    // Numbers, booleans, null, undefined — skip
  }

  /**
   * Check whether a keyword appears in the normalised text.
   * Matches whole words and hyphenated compound terms.
   * Uses a simple word-boundary check without regex to stay lightweight.
   */
  private keywordMatches(normalizedText: string, keyword: string): boolean {
    // For compound keywords (e.g. "blood-test"), replace hyphens with spaces
    // and also check hyphenated form, so both "blood test" and "blood-test" match.
    const spaced = keyword.replace(/-/g, " ");
    return (
      this.containsWholeWord(normalizedText, keyword) ||
      (spaced !== keyword && this.containsWholeWord(normalizedText, spaced))
    );
  }

  /**
   * Check that a `term` appears in `text` as a whole word (not mid-word).
   * Word characters are `[a-z0-9]`. Hyphens are treated as word characters
   * for compound terms.
   */
  private containsWholeWord(text: string, term: string): boolean {
    const index = text.indexOf(term);
    if (index === -1) return false;

    const before = index > 0 ? text[index - 1] : " ";
    const after =
      index + term.length < text.length ? text[index + term.length] : " ";

    const isWordChar = (char: string): boolean =>
      /[a-z0-9\-_]/.test(char);

    return !isWordChar(before ?? "") && !isWordChar(after ?? "");
  }

  /**
   * Count the total number of unique keywords defined across all rules
   * for a given domain.
   */
  private countKeywordsForDomain(domain: string): number {
    const keywords = new Set<string>();
    for (const rule of this.rules) {
      if (rule.domain === domain) {
        for (const kw of rule.keywords) {
          keywords.add(kw.toLowerCase());
        }
      }
    }
    return keywords.size;
  }
}
